enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,348 @@
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The current CLI login process involves multiple interactive steps where the user has to enter credentials multiple
+times. This process can be simplified so that the `oc` tool can be used to login directly without having to first login
+to the cluster console, the OAuth server and finally copying the login command. The `oc` tool can be modified to act as
+a public client to the Openshift OAuth server and received an access token
+through [Authorization Grant](https://datatracker.ietf.org/doc/html/rfc6749#section-1.3) flow
+and [PKCE](https://datatracker.ietf.org/doc/html/rfc7636).
+
+## Motivation
+
+Currently, when logging in with the `oc` client a user has to perform the following steps:
+
+1. Open the console in the browser.
+2. Login to console with the configured identity provider.
+3. Click on the **"Copy Login Command"** button which redirects the page to the OAuth application.
+4. Perform login again like in step 2.
+5. Copy the displayed login command.
+6. Run the copied command to configure the client to configure the client and access the cluster.
+
+This process can be simplified to the following steps:
+
+1. User runs `oc login --auto $CLUSTER_API_URL`
+2. `oc` starts a webserver to receive a callback from the OAuth server. It then opens the authorization endpoint with a
+   PKCE code challenge.
+3. The OAuth application performs the authentication either directly or by redirecting the user.
+4. Once authentication is successful the OAuth application redirects the user back to the callback URL of `oc` along
+   with a code.
+5. `oc` can then use this code along with the code challenge to request an access token. On receiving the token `oc`
+   then performs any configuration required to use the token.
+
+Using the above flow the user has to perform a login only once and the `oc` tool is configured automatically without any
+user input.
+
+### Goals
+
+Provide a one-step login for users which can be specified by the cluster operator.
+
+### Non-Goals
+
+TBD
+
+## Proposal
+
+### User Stories
+
+TBD
+
+### Implementation Details [optional]
+
+#### _oc login_ command
+
+The oc login command will have new 2 new flags, `--auto` and `--oauth-url`. The first flag indicates to `oc` that it
+should use the Oauth2 Authorization Flow to login. The second flag is optional and should be specified when the domain
+of the internal OAuth server has
+been [customized](https://docs.openshift.com/container-platform/4.9/authentication/configuring-internal-oauth.html#customizing-the-oauth-server-url_configuring-internal-oauth)
+. It will be the URL of the modified OAuth server.
```

Comment:
also one thing is that we should whether a browser is available - imagine if you're on a headless server in which case we should tell the user it's not possible to login that way.

here's how you can do it in python:
https://github.com/Azure/azure-cli/blob/c6afbd09efe35cebd72beeb47ba6d1893d28434e/src/azure-cli-core/azure/cli/core/util.py#L693

they have some funny hacks around detecting wsl and such, probably also useful for us to look into...

Reply:
which also brings up the question on how to even open a browser, there seems to be one popular framework: https://github.com/pkg/browser

Reply:
The way I've seen it work is that when there is no browser available the CLI just prints the URL and says open this. And people trying to run this on a server should recognize that the callback wont work. They could copy the URL and do the login on a machine with the browser and when they are done copy the callback URL and call that on the headless machine.

Reply:
> And people trying to run this on a server should recognize that the callback wont work  

IMHO it should fail and just tell you the option isn't available rather than misleading you that opening it in your browser and copying a curl back into your shell. How would that even work? The browser will display an empty page because there is nothing listening on the port and the post body with the credentials is lost.

@Anandnatraj wdyt?

Reply:
Callback's a GET with the token as parameter AFAIK.

Reply:
I only vaguely remember my time working with keycloak on things like that, but just checking their docs it is a redirect with parameters. So the copy&paste would probably work if people know what they're doing.